internal/model/modelindexer/indexer.go

```diff
@@ -434,6 +397,72 @@ func (i *Indexer) flush(ctx context.Context, bulkIndexer *bulkIndexer) error {
 	return nil
 }
 
+// runActiveIndexer starts a new active indexer which pulls items from the
+// bulkItems channel. The more active indexers there are, the faster events
+// will be pulled out of the queue, but also the more likely it is that the
+// outgoing Elasticsearch bulk requests are flushed due to the idle timer,
+// rather than due to being full.
+func (i *Indexer) runActiveIndexer() {
+	i.g.Go(func() error {
+		atomic.AddInt64(&i.activeBulkRequests, 1)
+		var mu sync.Mutex
+		// `active` and `timer` must be accessed with `mu` locked.
+		var active *bulkIndexer
+		var timer *time.Timer
+		flush := make(chan *bulkIndexer)
+		defer func() {
+			atomic.AddInt64(&i.activeBulkRequests, -1)
+			close(flush)
+		}()
+		for item := range i.bulkItems {
```

Comment:
I wonder if this could be simplified, and remove some locking, by selecting on the `flush` event in the foreground? Something like this:

```go
var closed bool
var active *bulkIndexer
flushTimer := time.NewTimer(0)
if !flushTimer.Stop() {
    <-flushTimer.C
}
for !closed {
    var shouldFlush bool
    select {
    case <-flushTimer.C:
        shouldFlush = true
    default:
        select {
        case <-flushTimer.C:
            shouldFlush = true
        case event, ok := <-i.bulkItems:
            if !ok {
                closed = true
                break
            }
            if active == nil {
                active = <-i.available
                atomic.AddInt64(&i.availableBulkRequests, -1)
                flushTimer.Reset(i.config.FlushInterval)
            }
            if err := active.Add(item); err != nil {
                i.logger.Errorf("failed adding event to bulk indexer: %v", err)
            }
            // Flush the active bulk indexer when it's at or exceeds the
            // configured FlushBytes threshold.
            if active.Len() >= i.config.FlushBytes {
                shouldFlush = true
                if !flushTimer.Stop() {
                    <-flushTimer.C
                }
            }
        }
    }

    if shouldFlush {
        indexer := active
        active = nil
        if !flushTimer.Stop() {
            <-flushTimer.C
        }
        i.g.Go(func() error {return i.flushIndexer(context.Background(), indexer)})
    }
}
```

Reply:
I've implemented a very similar approach in [5e5b24d](https://github.com/elastic/apm-server/pull/9318/commits/5e5b24da3f6c3f69ed83667cac3798a43f6d14c0). Let me know what you think.

Reply:
Looks great!